src/tacacs/pam/0007-Fix-memory-leak-when-parse-configuration.patch

```diff
@@ -0,0 +1,314 @@
++        }
++
+         return PAM_AUTH_ERR;
+     }
+ 
+@@ -664,7 +672,6 @@ int pam_sm_acct_mgmt (pam_handle_t * pamh, int flags,
+         free (arep.msg);
+ 
+     close(tac_fd);
+-
+     return status;
+ }    /* pam_sm_acct_mgmt */
+ 
+@@ -726,8 +733,9 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+ 
+     if (   (pam_get_item(pamh, PAM_OLDAUTHTOK, &pam_pass) == PAM_SUCCESS)
+         && (pam_pass != NULL) ) {
+-         if ((pass = strdup(pam_pass)) == NULL)
++         if ((pass = strdup(pam_pass)) == NULL) {
+               return PAM_BUF_ERR;
++         }
+     } else {
+         pass = strdup("");
+     }
+@@ -736,6 +744,7 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+         if(pass) {
+                 free(pass);
+         }
++
+         return PAM_USER_UNKNOWN;
+     }
+     
+@@ -762,7 +771,7 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+         if (ctrl & PAM_TAC_DEBUG)
+             syslog(LOG_DEBUG, "%s: trying srv %d", __FUNCTION__, srv_i );
+ 
+-        tac_fd = tac_connect_single(tac_srv[srv_i].addr, tac_srv[srv_i].key, tac_source_addr, tac_timeout, __vrfname);
++        tac_fd = tac_connect_single(tac_srv[srv_i].addr, tac_srv[srv_i].key, &tac_source_addr, tac_timeout, __vrfname);
+         if (tac_fd < 0) {
+             _pam_log(LOG_ERR, "connection failed srv %d: %m", srv_i);
+             continue;
+@@ -820,7 +829,8 @@ int pam_sm_chauthtok(pam_handle_t * pamh, int flags,
+                     communicating = 0;
+ 
+                     active_server.addr = tac_srv[srv_i].addr;
+-                    active_server.key = tac_srv[srv_i].key;
++                    /* copy secret to key */
++                    snprintf(active_server.key, sizeof(active_server.key), "%s", tac_srv[srv_i].key);
+ 
+                     if (ctrl & PAM_TAC_DEBUG)
+                         syslog(LOG_DEBUG, "%s: active srv %d", __FUNCTION__, srv_i);
+diff --git a/support.c b/support.c
+index be1f21b..3d38f24 100644
+--- a/support.c
++++ b/support.c
+@@ -30,7 +30,12 @@
+ #include <stdlib.h>
+ #include <string.h>
+ 
++/* tacacs server information */
+ tacplus_server_t tac_srv[TAC_PLUS_MAXSERVERS];
++struct addrinfo tac_srv_addr[TAC_PLUS_MAXSERVERS];
++struct sockaddr tac_sock_addr[TAC_PLUS_MAXSERVERS];
++struct sockaddr_in6 tac_sock6_addr[TAC_PLUS_MAXSERVERS];
++
+ int tac_srv_no = 0;
+ 
+ char tac_service[64];
+@@ -38,7 +43,11 @@ char tac_protocol[64];
+ char tac_prompt[64];
+ char *__vrfname=NULL;
+ char tac_source_ip[64];
+-struct addrinfo *tac_source_addr = NULL;
++
++/* source address */
++struct addrinfo tac_source_addr;
++struct sockaddr tac_source_sock_addr;
++struct sockaddr_in6 tac_source_sock6_addr;
+ 
+ void _pam_log(int err, const char *format,...) {
+     char msg[256];
+@@ -173,9 +182,38 @@ int tacacs_get_password (pam_handle_t * pamh, int flags
+     return PAM_SUCCESS;
+ }
+ 
++}
++/* set source ip address for the outgoing tacacs packets */
++void set_source_ip(const char *tac_source_ip) {
```

Comment:
>set_source_ip [](http://example.com/codeflow?start=6&length=13)

My understanding is that you are fixing a bug in 0006 patch, not the upstream. If true, please modify 0006 patch instead of patch-on-patch.

Reply:
This fix bugs in both 0006 patch and upstream.